cloudify_agent/worker.py

```diff
@@ -176,15 +194,73 @@ def new_excepthook(exception_type, value, the_traceback):
     sys.excepthook = new_excepthook
 
 
+class _ProcessRegistry(object):
+    """A registry for dispatch subprocesses.
+
+    The dispatch TaskHandler uses this to register the subprocesses that
+    are running and executing a task, so that they can be cancelled/killed
+    from outside.
+    """
+    def __init__(self):
+        self._processes = {}
+
+    def register(self, handler, process):
+        self._processes[self.make_key(handler)] = process
+
+    def unregister(self, handler):
+        self._processes.pop(self.make_key(handler), None)
+
+    def cancel(self, operation_id):
+        p = self._processes.get(operation_id)
+        if p:
+            t = threading.Thread(target=self._stop_process, args=(p, ))
+            t.start()
+
+    def _stop_process(self, process):
+        """Stop the process: SIGTERM, and after 5 seconds, SIGKILL
+
+        Note that on windows, both terminate and kill are effectively
+        the same operation."""
+        process.terminate()
+        for i in range(10):
+            if process.poll() is not None:
+                return
+            time.sleep(0.5)
+        process.kill()
+
+    def make_key(self):
+        raise NotImplementedError()
+
+
+class OperationProcessRegistry(_ProcessRegistry):
+    def make_key(self, handler):
+        return handler.ctx.task_id
+
+
+class WorkflowProcessRegistry(_ProcessRegistry):
+    def make_key(self, handler):
+        return handler.ctx.execution_id
+
+    def _stop_process(self, process):
+        time.sleep(10)
+        super(WorkflowProcessRegistry, self)._stop_process(process)
+
+
 def make_amqp_worker(args):
     if args.name:
         _setup_excepthook(args.name)
     _setup_logger(args.name)
 
+    operation_registry = OperationProcessRegistry()
```

Comment:
Is there a reason why those aren't global? Seems a bit weird to create them here and then to pass them to all of the consumers. I think you maybe have already explained this, but I forgot.

Reply:
i do prefer dependency injection (even manual, without a framework) to globals